llama.cpp

```diff
@@ -2112,6 +2280,32 @@ void llama_sample_frequency_and_presence_penalties(struct llama_context * ctx, l
     }
 }
 
+void llama_sample_grammar(struct llama_context * ctx, llama_token_data_array * candidates, const struct llama_grammar * grammar) {
+    assert(ctx);
+    const int64_t     t_start_sample_us  = ggml_time_us();
+    const llama_token eos                = llama_token_eos();
+    // since many llama tokens are prefixed with a single space, special case a lookahead on ' '
+    const auto        stacks_after_space = llama_grammar_accept(grammar->rules, grammar->stacks, ' ');
+
+    for (size_t i = 0; i < candidates->size; ++i) {
+        const llama_token id    = candidates->data[i].id;
+        const char *      str   = llama_token_to_str(ctx, id);
+
+        // prune tokens based on first char only - in `llama_grammar_accept_token` we will find the
+        // full matching prefix of the selected token
+        const bool valid = str[0] == ' '
+            ? llama_grammar_peek(stacks_after_space, str[1])
+            : str[0] || id == eos
+            ? llama_grammar_peek(grammar->stacks,    id == eos ? 0 : str[0])
+            : false;
```

Comment:
Is it possible to rewrite this to be clearer without the ternary ops?

Something like this, if it is indeed how I understood it?

```suggestion
        bool valid = false;
        if (id == eos) {
            valid = llama_grammar_peek(grammar->stacks, 0);
        } else if (str[0] == ' ') {
            valid = llama_grammar_peek(stacks_after_space, str[1]);
        } else if (str[0] != 0) {
            valid = llama_grammar_peek(grammar->stacks, str[0]);
        }
```

Reply:
Yes, that's better, thanks!